### PR TITLE
bug: ProverSync from_disk now properly recovers tree state, proofs, etc.

### PR DESCRIPTION
**Pull request by prestwich**
- There was a significant bug where `from_disk` would cause `incremental` and `prover` to have different states
- while fixing this, I did some significant refactors to prover sync to have it rely on HomeIndexer and not block
- also made it run a proof-presence check on startup to ensure that we don't somehow miss proofs

- To consider: ProverSync no longer makes RPC requests so it is 100% sync now. What's a reasonable way (besides `wait_for_leaf`) to make it yield priority to other tasks?

### CODE REVIEW

**Comment by emberian** [comment]
> To consider: ProverSync no longer makes RPC requests so it is 100% sync now. What's a reasonable way (besides wait_for_leaf) to make it yield priority to other tasks?

don't! task scheduling is the responsibility of the execution context. use `tokio::task::spawn_blocking`, and it will take care of the rest.

File: rust/optics-core/src/db/mod.rs [hunk 1]
```diff
@@ -257,32 +307,20 @@ impl DB {
...
-        let mut path = Path::new(".").canonicalize()?;
-        path.extend(&[db_path]);
-
-        match path.is_dir() {
-            true => info!(
-                "Opening existing db at {path}",
-                path = path.to_str().unwrap()
-            ),
-            false => info!("Creating db at {path}", path = path.to_str().unwrap()),
+    // TODO(james): this is a quick-fix for the prover_sync and I don't like it
```
Comment (emberian) [inline rust/optics-core/src/db/mod.rs:313]: eventually consistent, inside the same process. i dig it.

File: rust/agents/processor/src/prover_sync.rs [hunk 2]
```diff
@@ -69,14 +67,67 @@ pub enum ProverSyncError {
...
-    pub fn new(prover: Prover, home: Arc<Homes>, db: DB) -> Self {
-        Self {
+    fn store_proof(&self, leaf_index: u32) -> Result<(), ProverSyncError> {
+        let proof = self.prover.prove(leaf_index as usize)?;
+        self.db.store_proof(leaf_index, &proof)?;
+        info!(
+            leaf_index,
+            root = ?self.prover.root(),
+            "Storing proof for leaf {}",
+            leaf_index
```
Comment (emberian) [inline rust/agents/processor/src/prover_sync.rs:77]: how do you feel about `"Storing proof", leaf_index = leaf_index` ? although i don't know that the short formatter includes fields...

File: rust/agents/processor/src/prover_sync.rs [hunk 3]
```diff
@@ -69,14 +67,67 @@ pub enum ProverSyncError {
...
+        let sync = Self {
             prover,
-            home,
-            incremental: IncrementalMerkle::default(),
+            incremental,
             db,
+        };
+
+        // Ensure proofs exist for all leaves
+        for i in 0.. {
```
Comment (emberian) [inline rust/agents/processor/src/prover_sync.rs:119]: 👍🏼 i love executable postconditions

**Review by emberian — APPROVED** [review]

File: rust/agents/processor/src/prover_sync.rs (hunk 2 above)
Comment (prestwich) [reply rust/agents/processor/src/prover_sync.rs:77]: ah you mean using named formatter args in the format string? I don't see a compelling reason to do it but I don't mind it

File: rust/agents/processor/src/prover_sync.rs (hunk 3 above)
Comment (prestwich) [reply rust/agents/processor/src/prover_sync.rs:119]: this made me think of an edge case I want to handle. will push up a commit in a sec

**Review by emberian — APPROVED** [review]

**Pull request merged**